File: nvpl_blas/f77/zgemm_batch_strided.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     ZGEMM_BATCH_STRIDED 
!
!******************************************************************************
      program   ZGEMM_BATCH_STRIDED_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

      character*1      transa, transb
      parameter        (transa='N', transb='N')
      integer          m, n, k, i, j, batch_size
      integer          lda, ldb, ldc
      parameter        (m=2, n=2, k=2, batch_size=2,
     &                  lda=2, ldb=2, ldc=2)
      integer          stridea, strideb, stridec
      parameter        (stridea=lda*k, strideb=ldb*n, stridec=ldc*n)
      double complex   alpha, beta
      parameter        (alpha=(1.0,2.0), beta=(1.0,2.0))
      double complex   a(lda,k,batch_size), b(ldb,n,batch_size),
     &                 c(ldc,n,batch_size)

*       External Subroutines
      external         print_zmatrix, fill_zmatrix
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, m, n, k, lda, ldb, ldc, transa, transb
      print 103, stridea, strideb, stridec, batch_size
      print 101, alpha, beta
      do i = 1, batch_size
          call fill_zmatrix(a(:,:,i),m,k,lda,'F','N')
          call fill_zmatrix(b(:,:,i),k,n,ldb,'F','N')
          call fill_zmatrix(c(:,:,i),m,n,ldc,'F','N')
          call print_zmatrix(a(:,:,i),m,k,lda,'A')
          call print_zmatrix(b(:,:,i),k,n,ldb,'B')
          call print_zmatrix(c(:,:,i),m,n,ldc,'C')
      end do
*
*      Call ZGEMM_BATCH_STRIDED subroutine
      call ZGEMM_BATCH_STRIDED(transa,transb,m,n,k,alpha,a,lda,
     &     stridea,b,ldb,strideb,beta,c,ldc,stridec,batch_size)
*
      print*
      print 102, 'After ZGEMM_BATCH_STRIDED operation:'
      do i = 1, batch_size
          call print_zmatrix(c(:,:,i),m,n,ldc,'C')
      end do

      stop
 99   format('Example: ZGEMM_BATCH_STRIDED for the matrix-matrix'
     &       ' multiplication of a batch of matrices')
 100  format('#### args: m=',i1,', n=',i1,', k=',i1,', lda=',i1,
     &       ', ldb=',i1,', ldc=',i1,', transA=',a1,', transB=',a1)
 101  format(11x,'alpha=(',f4.2,', ',f4.2,'),'
     &       ' beta=(',f4.2,', ',f4.2,')')
 102  format(A)
 103  format(11x,'stridea=',i1,', strideb=',i1,', stridec=',i1,
     &       ', batch_size=',i1)
      end

File: nvpl_blas/c/zgemm_batch_strided.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_zgemm_batch_strided
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t M = 2;
    nvpl_int_t N = 2;
    nvpl_int_t K = 2;
    nvpl_int_t lda = 2;
    nvpl_int_t ldb = 2;
    nvpl_int_t ldc = 2;
    nvpl_dcomplex_t alpha = {1.0f, 2.0f};
    nvpl_dcomplex_t beta = {1.0f, 2.0f};
    nvpl_int_t batch_size = 2;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_TRANSPOSE transA = CblasNoTrans;
    enum CBLAS_TRANSPOSE transB = CblasNoTrans;
    nvpl_dcomplex_t * A = NULL;
    nvpl_dcomplex_t * B = NULL;
    nvpl_dcomplex_t * C = NULL;
    int64_t matrixSizeA = 0;
    int64_t matrixSizeB = 0;
    int64_t matrixSizeC = 0;

    nvpl_int_t rowsA = (transA == CblasNoTrans) ? M : K;
    nvpl_int_t colsA = (transA == CblasNoTrans) ? K : M;
    nvpl_int_t rowsB = (transB == CblasNoTrans) ? K : N;
    nvpl_int_t colsB = (transB == CblasNoTrans) ? N : K;
    nvpl_int_t rowsC = M;
    nvpl_int_t colsC = N;

    matrixSizeA = (int64_t)lda * colsA;
    matrixSizeB = (int64_t)ldb * colsB;
    matrixSizeC = (int64_t)ldc * colsC;
    nvpl_int_t strideA = matrixSizeA;
    nvpl_int_t strideB = matrixSizeB;
    nvpl_int_t strideC = matrixSizeC;
    printf("\nExample: cblas_zgemm_batch_strided for the matrix-matrix multiplication of a batch of matrices\n\n");
    printf("#### args: m=%" PRId64 ", n=%" PRId64 ", k=%" PRId64 ", lda=%" PRId64 ", ldb=%" PRId64 ", ldc=%" PRId64 ", "
            "transA=%c, transB=%c, order=%c\n", (int64_t)M, (int64_t)N, (int64_t)K, (int64_t)lda, (int64_t)ldb,
            (int64_t)ldc, transpose_to_char(transA), transpose_to_char(transB), order_to_char(order));
    printf("           strideA=%" PRId64 ", strideB=%" PRId64 ", strideC=%" PRId64 ", batch_size=%" PRId64 "\n",
            (int64_t)strideA, (int64_t)strideB, (int64_t)strideC, (int64_t)batch_size);
    printf("           alpha=(%g, %g), beta=(%g, %g)\n", alpha.real, alpha.imag, beta.real, beta.imag);

    // allocate memory
    A = (nvpl_dcomplex_t *)malloc((strideA * (int64_t)batch_size) * sizeof(nvpl_dcomplex_t));
    B = (nvpl_dcomplex_t *)malloc((strideB * (int64_t)batch_size) * sizeof(nvpl_dcomplex_t));
    C = (nvpl_dcomplex_t *)malloc((strideC * (int64_t)batch_size) * sizeof(nvpl_dcomplex_t));

    // fill data
    for (nvpl_int_t i = 0; i < batch_size; ++i) {
        fill_zmatrix(A + strideA * i, rowsA, colsA, lda, order, Full, CblasNonUnit);
        fill_zmatrix(B + strideB * i, rowsB, colsB, ldb, order, Full, CblasNonUnit);
        fill_zmatrix(C + strideC * i, rowsC, colsC, ldc, order, Full, CblasNonUnit);
    }

    // print input data
    for (nvpl_int_t i = 0; i < batch_size; ++i) {
        print_zmatrix(A + strideA * i, rowsA, colsA, lda, "A", order);
        print_zmatrix(B + strideB * i, rowsB, colsB, ldb, "B", order);
        print_zmatrix(C + strideC * i, rowsC, colsC, ldc, "C", order);
    }

    // call cblas_zgemm_batch_strided
    cblas_zgemm_batch_strided(order, transA, transB, M, N, K, &alpha, A, lda, strideA,
            B, ldb, strideB, &beta, C, ldc, strideC, batch_size);

    // print result
    printf("\nAfter gemm operation:\n");
    for (nvpl_int_t i = 0; i < batch_size; ++i) {
        print_zmatrix(C + strideC * i, rowsC, colsC, ldc, "C", order);
    }


    // release memory
    free(A);
    free(B);
    free(C);
    return 0;
}
